### PR TITLE
Add request and response messages.

### PR DESCRIPTION
**Pull request by TheJMaster**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -5,15 +5,23 @@ use std::thread;
...
+const BUFFER_SIZE: usize = 256;
 const PORT: u16 = 3333;
 
 fn handle_client(mut stream: TcpStream) {
     let mut data = [0 as u8; BUFFER_SIZE];
     while match stream.read(&mut data) {
         Ok(size) => {
             trace!("stream read {} bytes", size);
-            stream.write(&data[0..size]).unwrap();
+            let request = crate::message::Request::deserialize(&data);
```
Comment (LukasJoswiak) [inline src/main.rs]: Why pass data by reference? You can probably just transfer ownership to the deserialize method.

Comment (LukasJoswiak) [inline src/main.rs]: What happens if the message doesn't match the format expected for deserialization?

File: src/main.rs [hunk 2]
```diff
@@ -5,15 +5,23 @@ use std::thread;
 use std::net::{TcpListener, TcpStream, Shutdown};
 use std::io::{Read, Write};
 
-const BUFFER_SIZE: usize = 250;
+pub mod translate;
+pub mod message;
```
Comment (LukasJoswiak) [inline src/main.rs:10]: These seem to be listed in logical order, but they should probably be listed in alphabetical order instead.

File: src/message.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Request {
+    pub id: usize,
+    pub lang: crate::translate::Language,
+    pub text: String,
+}
+
+impl Request {
+    pub fn serialize(&self) -> Vec<u8> {
+        serde_json::to_vec(&self).unwrap()
```
Comment (LukasJoswiak) [inline src/message.rs:13]: Add error handling if serialization fails.

File: src/message.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    pub fn deserialize(serialized: &[u8]) -> Request {
+        serde_json::from_slice(&serialized).unwrap()
```
Comment (LukasJoswiak) [inline src/message.rs]: Add error handling (logging) if deserialization fails.

File: src/message.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Response {
+    pub id: usize,
+    pub text: String,
+}
+
+impl Response {
+    pub fn serialize(&self) -> Vec<u8> {
+        serde_json::to_vec(&self).unwrap()
```
Comment (LukasJoswiak) [inline src/message.rs:29]: See comments above.

File: src/message.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    pub fn deserialize(serialized: &[u8]) -> Response {
+        serde_json::from_slice(&serialized).unwrap()
```
Comment (LukasJoswiak) [inline src/message.rs]: See comments above.

**Review by LukasJoswiak — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (TheJMaster) [reply src/main.rs]: Discussed in person, needs to be a reference since array size it unknown.

**Review by LukasJoswiak — APPROVED** [review]

**Pull request merged**